### PR TITLE
feat(ODC-451): add function that checks if repo is reachable and branch exists

### PR DESCRIPTION
**Pull request by MatousJobanek**
the `IsReachableWithBranch` function checks whether the repo is reachable or not. If it is (which means that the URL is correct and repo is public), then it checks if the provided branch exists
it is needed for this: https://github.com/redhat-developer/devconsole-operator/pull/118

### CODE REVIEW

File: pkg/git/repository/connectivity.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	if !strings.HasSuffix(path, ".git") {
+		path += ".git"
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	url := fmt.Sprintf("https://%s%s/info/refs?service=git-upload-pack", endpoint.Host, path)
+	resp, err := client.Get(url)
+	if err != nil || endpoint.Host == "" || endpoint.Path == "" {
+		return false, fmt.Errorf(unableReachRepoError)
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/git/repository/connectivity.go]: yes, or more likely fix the wrong if statement - fixed 9169c26

File: pkg/git/repository/connectivity.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	err = resp.Body.Close()
+	if err != nil {
+		log.Error(err, "error while closing body")
+		return false, fmt.Errorf(unableReachRepoError)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf(unableReachRepoError)
+	} else {
```
Comment (tinakurian) [inline pkg/git/repository/connectivity.go]: can we please remove the else statement here? it doesn't look like it's necessary :-)

File: pkg/git/repository/connectivity.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		if branch == "" {
+			branch = Master
+		}
+		branchRefRegexp := fmt.Sprintf(`\ refs\/heads\/%s\n`, branch)
+		compile, err := regexp.Compile(branchRefRegexp)
+		if err != nil {
+			log.Error(err, "error while parsing regexp of the given branch")
+			return false, fmt.Errorf(unableParseBranchError)
+		} else {
```
Comment (tinakurian) [inline pkg/git/repository/connectivity.go]: can we please remove the else clause here? we can probably just have 
```
if compile.Find(body) != nil { 
     return true, nil
}

return false, fmt.Errorf(unableFindBranchError)
```

File: pkg/git/repository/connectivity.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	"strings"
+)
+
+const (
+	unableReachRepoError   = "unable to reach the repo"
+	unableFindBranchError  = "unable to find the branch"
+	unableParseBranchError = "unable to parse the branch"
+)
+
+func IsReachableWithBranch(log *log.GitSourceLogger, branch string, endpoint *gittransport.Endpoint) (bool, error) {
```
Comment (tinakurian) [inline pkg/git/repository/connectivity.go]: could we please add a func comment here describing what the function does?

File: pkg/git/repository/connectivity.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		path += ".git"
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	url := fmt.Sprintf("https://%s%s/info/refs?service=git-upload-pack", endpoint.Host, path)
+	resp, err := client.Get(url)
+	if err != nil {
+		return false, fmt.Errorf(unableReachRepoError)
+	} else {
```
Comment (tinakurian) [inline pkg/git/repository/connectivity.go]: can we remove this else clause and just have 

`return validateBranch(log, branch, resp)`

File: pkg/git/repository/connectivity.go (hunk 2 above)
Comment (MatousJobanek) [reply pkg/git/repository/connectivity.go]: yeah, good point - it's a zombie after refactoring of the final code 227e92b

File: pkg/git/repository/connectivity.go (hunk 3 above)
Comment (MatousJobanek) [reply pkg/git/repository/connectivity.go]: yup, done 227e92b

File: pkg/git/repository/connectivity.go (hunk 5 above)
Comment (MatousJobanek) [reply pkg/git/repository/connectivity.go]: done 227e92b

File: pkg/git/repository/connectivity.go (hunk 4 above)
Comment (MatousJobanek) [reply pkg/git/repository/connectivity.go]: added here 919a1a0

**Comment by MatousJobanek** [comment]
@tinakurian I changed it a bit more as I incorporated also the changes from [devconsole-api](https://github.com/redhat-developer/devconsole-api/commit/36791a68b58dd10ed2ead21974bb9101b4ebe51e) repo and aligned with the new validations - see: https://github.com/redhat-developer/devconsole-git/pull/26. All your comments should be addressed

**Pull request merged**